datafusion/expr-common/src/signature.rs

```diff
@@ -286,6 +261,34 @@ impl Display for ArrayFunctionSignature {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Hash)]
+pub enum ArrayFunctionArgument {
+    /// A non-list or list argument. The list dimensions should be one less than the Array's list
+    /// dimensions.
+    Element,
+    /// An Int64 index argument.
+    Index,
```

Comment:
Offset might be a better name for this? It can technically be used for sizes in functions like `array_resize` or counts for functions like `array_replace_n`.